DB schema extension for Debian10-Minions on cpuarch=aarch64,osarch=arm64, Issue #2548

Shirocco88 opened the pull request:
## What does this PR change?
DB schema extension for Debian10-Minions on cpuarch=aarch64,osarch=arm64, Issue #2548

        modified:   schema/spacewalk/common/data/rhnChannelArch.sql
        modified:   schema/spacewalk/common/data/rhnChannelPackageArchCompat.sql
        modified:   schema/spacewalk/common/data/rhnChildChannelArchCompat.sql
        modified:   schema/spacewalk/common/data/rhnCpuArch.sql
        modified:   schema/spacewalk/common/data/rhnPackageArch.sql
        modified:   schema/spacewalk/common/data/rhnPackageUpgradeArchCompat.sql
        modified:   schema/spacewalk/common/data/rhnServerArch.sql
        modified:   schema/spacewalk/common/data/rhnServerChannelArchCompat.sql
        modified:   schema/spacewalk/common/data/rhnServerPackageArchCompat.sql
        modified:   schema/spacewalk/common/data/rhnServerServerGroupArchCompat.sql
        modified:   schema/spacewalk/susemanager-schema.changes
        new file:   schema/spacewalk/upgrade/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/120-arm64.sql

## GUI diff

No difference.

Before:

After:

- [ ] **DONE**

## Documentation
- No documentation needed: **add explanation. This can't be used if there is a GUI diff**
- Documentation issue was created: [Link for SUSE Manager contributors](https://github.com/SUSE/spacewalk/issues/new?template=ISSUE_TEMPLATE_DOCUMENTATION.md&labels=documentation&projects=SUSE/spacewalk/31), [Link for community contributors](https://github.com/uyuni-project/uyuni-docs/issues/new).
- (OPTIONAL) [Documentation PR](https://github.com/uyuni-project/uyuni-docs/pulls)

- [ ] **DONE**

## Test coverage
- No tests: **add explanation**
- Unit tests were added
- Cucumber tests were added

- [ ] **DONE**

## Links
Issue #2548
Tracks # **add downstream PR, if any**

- [ ] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)


## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

Shirocco88 commented:
@Uyuni-Team: 
there seams to be a problem with this PR und check "schema_migration_test_pgsql"
unfortunately I can't view to output of this test;

could someone post the output of pipeline: 
https://ci.suse.de/job/uyuni-prs-schema_migration_test_pgsql-pipeline/52702/

so that I can check and fix the problem;

Best regards
Shirocco88

juliogonzalez commented:
```
                                   ?column?                                    
-------------------------------------------------------------------------------
 susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/120-arm64.sql.postgresql
(1 row)

INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
psql:<stdin>:99: ERROR:  syntax error at or near "all"
LINE 6: LOOKUP_PACKAGE_ARCH('all-deb'), LOOKUP_PACKAGE_ARCH('arm64-d...
```

Looks to me that the function `LOOKUP_PACKAGE_ARCH` does not know what to do with `all-deb`.

Shirocco88 commented:
@juliogonzalez: Thanks
I'll will check and fix this with psql-client;

Shirocco88 commented:
@juliogonzalez: 
I've duplicated the definition of amd86 to arm86;

But one insert statement causes an error:
I've marked this statement as comment in the sql-file:

```
schema/spacewalk/upgrade/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/120-arm64.sql
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('amd64-deb'), LOOKUP_PACKAGE_ARCH('all-deb'), current_timestamp, current_timestamp);
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('amd64-deb'), LOOKUP_PACKAGE_ARCH('amd64-deb'), current_timestamp, current_timestamp);
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('all-deb'), LOOKUP_PACKAGE_ARCH('amd64-deb'), current_timestamp, current_timestamp);
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('arm64-deb'), LOOKUP_PACKAGE_ARCH('all-deb'), current_timestamp, current_timestamp);
-- insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('arm64-deb'), LOOKUP_PACKAGE_ARCH('arm64-deb'), current_timestamp, current_timestamp);
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('all-deb'), LOOKUP_PACKAGE_ARCH('arm64-deb'), current_timestamp, current_timestamp);
```

```
schema/spacewalk/common/data/rhnPackageUpgradeArchCompat.sql
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('amd64-deb'), LOOKUP_PACKAGE_ARCH('all-deb'), current_timestamp, current_timestamp);
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('amd64-deb'), LOOKUP_PACKAGE_ARCH('amd64-deb'), current_timestamp, current_timestamp);
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('all-deb'), LOOKUP_PACKAGE_ARCH('amd64-deb'), current_timestamp, current_timestamp);
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('arm64-deb'), LOOKUP_PACKAGE_ARCH('all-deb'), current_timestamp, current_timestamp);
-- insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('arm64-deb'), LOOKUP_PACKAGE_ARCH('arm64-deb'), current_timestamp, current_timestamp);
insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('all-deb'), LOOKUP_PACKAGE_ARCH('arm64-deb'), current_timestamp, current_timestamp);
```

Please assign a reviewer/reviewers to this PR;

Thanks in advance;

Best regards
Shirocco88

juliogonzalez commented on schema/spacewalk/common/data/rhnPackageUpgradeArchCompat.sql:
```diff
@@ -144,4 +144,9 @@ insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_i
 insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('amd64-deb'), LOOKUP_PACKAGE_ARCH('amd64-deb'), current_timestamp, current_timestamp);
 insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('all-deb'), LOOKUP_PACKAGE_ARCH('amd64-deb'), current_timestamp, current_timestamp);
+insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('arm64-deb'), LOOKUP_PACKAGE_ARCH('all-deb'), current_timestamp, current_timestamp);
+-- insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) values (LOOKUP_PACKAGE_ARCH('arm64-deb'), LOOKUP_PACKAGE_ARCH('arm64-deb'), current_timestamp, current_timestamp);
```
I don't really understand, why do we need it commented?

Shirocco88 replied:
09/02/2020:
Bug in schema/spacewalk/upgrade/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/120-arm64.sql  
found:    missing apostroph in "and package_upgrade_arch_id = LOOKUP_PACKAGE_ARCH(arm64-deb'));"

problem fixed und new commit / force-push done;

juliogonzalez commented on schema/spacewalk/upgrade/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1/120-arm64.sql:
```diff
@@ -0,0 +1,96 @@
+LOOKUP_PACKAGE_ARCH('arm64-deb'), LOOKUP_PACKAGE_ARCH('all-deb'), current_timestamp, current_timestamp from dual
+where not exists (select 1 from rhnPackageUpgradeArchCompat where package_arch_id = LOOKUP_PACKAGE_ARCH('arm64-deb') and package_upgrade_arch_id = LOOKUP_PACKAGE_ARCH('all-deb'));
+
+-- insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) select 
```
Same: why do we need it commented?

Shirocco88 replied:
I would say, these statements are obsolet;
dito the corresponding statements for "amd64"

```
uyuni=# insert into rhnPackageUpgradeArchCompat (package_arch_id, package_upgrade_arch_id, created, modified) select
LOOKUP_PACKAGE_ARCH('arm64-deb'), LOOKUP_PACKAGE_ARCH('arm64-deb'), current_timestamp, current_timestamp from dual
where not exists (select 1 from rhnPackageUpgradeArchCompat where package_arch_id = LOOKUP_PACKAGE_ARCH('arm64-deb') and package_upgrade_arch_id = LOOKUP_PACKAGE_ARCH(arm64-deb'));
uyuni'#
```

Because of the foreign key constraints on table rhnPackageUpgradeArchCompat:
https://github.com/uyuni-project/uyuni/blob/master/schema/spacewalk/common/tables/rhnPackageUpgradeArchCompat.sql
it is not possible to insert the same value twice;

Shirocco88 commented:
@juliogonzalez: I'll remove the inactive statements

Shirocco88 commented:
@juliogonzalez: 
I've duplicated the DB schema definition from 'amd86' to 'arm86';
problem has been fixed;
all checks/tests are OK;

what has to be done in order to merge this PR ?

Thanks in advance;

Best regards
Shirocco88

juliogonzalez commented:
A reviewer is needed (I don't have enough knowledge of the internals of the DB for this PR), so I pinged people and there should be a reviewer soon.

cbbayburt commented:
@Shirocco88 thanks for the patch! We've put it in our backlog to review it as soon as possible.
Meanwhile, have you tested it manually with onboarding?

Shirocco88 commented:
@cbbayburt 
>> Meanwhile, have you tested it manually with onboarding?

Yes, there are also problems with "Java code", e.g. HardwareMapper.java;
I've opened a separate issue https://github.com/uyuni-project/uyuni/issues/2579

>> We've put it in our backlog to review it as soon as possible.

Looking forward the DB schema extension is done.

mcalmer reviewed (approved):
LGTM - the changelog has a conflict and we are about to release a new uyuni version.
If this PR cannot get merged before the release, you need to change the update script and move it
into the directory for the next update.

juliogonzalez commented:
> If this PR cannot get merged before the release, you need to change the update script and move it into the directory for the next update.

In fact that needs to happen already, as I tagged 4.2.1 already. I announced it at the uyuni-devel mailing list.

@Shirocco88, please rebase changes from master, and move the migration script to `susemanager-schema-4.2.1-to-susemanager-schema-4.2.2/`

juliogonzalez commented:
If you can handle it today, I will make sure it's part of Uyuni 2020.09

Shirocco88 commented:
@juliogonzalez: I'll will update the PR
thanks in advance!!!

Shirocco88 commented:
juliogonzalez:
> please rebase changes from master, and move the migration script to susemanager-schema-4.2.1-to-susemanager-schema-4.2.2/
...
>> If you can handle it today, I will make sure it's part of Uyuni 2020.09

@juliogonzalez: moved the migration script to susemanager-schema-4.2.1-to-susemanager-schema-4.2.2/
Updated changelog and PR

thanks in advance!!!

Shirocco88 commented:
Did "git rebase master" and add changelog

Looking forward to see the DB schema extension in Uyuni 2020.09.

juliogonzalez commented on schema/spacewalk/susemanager-schema.changes:
```diff
@@ -1,4 +1,9 @@
+Mon Sep 21 14:00:22 CEST 2020 - shirocco88
+
+- version 4.2.1-2
+- Schema upgrade for Debian with ARM64v8+ (cpuach=aarch64, cpuarch=arm64) architecture
```
No, this is not right.

Just add your entry after on top of the file, then an empty line.

The headers and the version change are added by the Release Engineer (me, in this case) with `tito tag`.

juliogonzalez replied:
Or in other words, you just need:
```
- Schema upgrade for Debian with ARM64v8+ (cpuach=aarch64, cpuarch=arm64) architecture

```
On top of what the file already has on master.

juliogonzalez replied:
Perfect!

juliogonzalez reviewed (changes requested):
A quick fix for the changelog, and we'll be ready :-)

Shirocco88 commented:
@juliogonzalez: changelog fixed
now PR should be ready for merging ;-)

juliogonzalez reviewed (approved):
LGTM :-)

juliogonzalez commented:
Ok, let's merge and tag.
